src/main/kotlin/org/rust/ide/annotator/RsSyntaxErrorsAnnotator.kt

```diff
@@ -321,30 +322,68 @@ private fun isComplexPattern(pat: RsPat?): Boolean {
     }
 }
 
-private fun checkValueParameter(holder: AnnotationHolder, param: RsValueParameter) {
-    val parent = param.parent.parent ?: return
-    when (parent) {
+private fun checkPat(holder: AnnotationHolder, element: RsPat) {
+    if (element is RsPatRange) {
+        checkPatRange(holder, element)
+    }
+
+    checkGeneralPat(element, holder)
+}
+
+private fun checkGeneralPat(element: RsPat, holder: AnnotationHolder) {
+    val valueParameter = element.parent
+    if (valueParameter !is RsValueParameter) {
+        return
+    }
+    val fn = valueParameter.parent.parent ?: return
+    when (fn) {
         is RsFunction -> {
-            checkValueParameterInFunction(parent, param, holder)
+            checkFunctionParameterPatIsSimple(fn, element, holder)
         }
 
         is RsFnPointerType -> {
-            checkValueParameterInFunctionPointerType(param, holder)
+            checkFunctionPointerTypeParameterPatIsSimple(element, holder)
         }
     }
 }
 
-fun checkValueParameterInFunctionPointerType(param: RsValueParameter, holder: AnnotationHolder) {
-    val pat = param.pat
-    val isComplexPattern = isComplexPattern(pat)
-    if (isComplexPattern) {
-        RsDiagnostic.PatternArgumentInFunctionPointerTypeError(param).addToHolder(holder)
+private fun checkFunctionParameterPatIsSimple(fn: RsFunction, element: RsPat, holder: AnnotationHolder) {
+    val isComplexPattern = isComplexPattern(element)
+    when (fn.owner) {
+        is RsAbstractableOwner.Foreign -> {
+            if (isComplexPattern) {
+                RsDiagnostic.PatternArgumentInForeignFunctionError(element).addToHolder(holder)
+            }
+        }
+
+        is RsAbstractableOwner.Trait -> {
+            if (isComplexPattern && fn.block == null) {
+                RsDiagnostic.PatternArgumentInFunctionWithoutBodyError(element).addToHolder(holder)
+            }
+        }
+
+        else -> {}
+    }
```

Comment:
I think after extraction from `checkValueParameterInFunction` is can be simplified:

```suggestion
    val owner = fn.owner
    val isComplexPatternForbidden = owner is RsAbstractableOwner.Foreign
        || owner is RsAbstractableOwner.Trait && fn.block == null

    if (isComplexPatternForbidden && isComplexPattern(element)) {
        RsDiagnostic.PatternArgumentInForeignFunctionError(element).addToHolder(holder)
    }
```

Reply:
But if `fn.owner` is `RsAbstractableOwner.Trait`, `RsDiagnostic.PatternArgumentInFunctionWithoutBodyError` is emitted, not `RsDiagnostic.PatternArgumentInForeignFunctionError`

Reply:
Aah, damn! Then nevermind :)